### DISCUSSION TITLE
Sound for py5

### DISCUSSION THREAD

**Discussion by jaivanti** [py5coding/py5generator#106] [Ideas]
Hello! I and Srushti have started understanding the works around implementing sound in py5. Would like to start the discussion here

**Comment by srukale** [comment]
hi @hx2A ! We had some question regarding sound in py5. Should we search for other packages for sound? Can we also implement using processing's sound library ?

**Reply by hx2A** [reply]
Hello, @srukale and @jaivanti !

The first step here is to start researching Python sound libraries. I know about simpleaudio but would be surprised if the were no other decent options. After doing some Googling I came across playsound, which looks interesting. There are others.

What functionality does Processing's sound library offer? That would be a good guide for what sound-related features users would find desirable.

While doing all of this, think about each library's dependencies and how hard they are to install and get working. Something with complicated dependencies & a painful setup process isn't going to be a good fit.

Next, test them out in a Sketch. How do you use them? How do you use them well? Does loading or playing audio cause any performance issues? How easy is it to control?

Once you have a good understanding of how to use a sound library with py5, we can think about next steps. I am averse to adding something to py5 with challenging dependencies to install because it will make it harder for beginners to get py5 up and running. If the sound library you find is decent and easy to explain, I think we'd just need a How-to page or two to explain to readers how to add sound to their Sketches.

**Reply by srukale** [reply]
Thank you , will start researching @hx2A

**Reply by srukale** [reply]
Hey @hx2A while exploring Processing's sound library , we came across [this  ](https://github.com/processing/processing-sound/tree/master/src/processing/sound). And we are thinking to implement something similar to this . Wanted to know your thoughts on this.

**Reply by hx2A** [reply]
That is the source code for Processing's sound library. It looks like it makes use of [JSyn](http://www.softsynth.com/jsyn/) for at least some of its functionality. Are you suggesting you'd like to create a new library that models its functionality after Processing's sound library?

The first thing you need to do is decide what Python libraries you want your library to be dependent on. Good choices will be easy to install without weird and complicated dependencies and they will be well maintained libraries.

Some Python libraries that would be useful here are numpy & scipy for reading wav audio files and working with the data.

The Python library [pydub](https://github.com/jiaaro/pydub) looks like a good choice for reading other audio formats, but the ffmpeg dependency might be a bit tricky for some users and some platforms. How easy it is to install on OSX or on Windows?

Same thing goes for the Python library [librosa](https://librosa.org/doc/latest/index.html), which also depends on ffmpeg. This looks like a good choice for doing beat detection and other fancy stuff. I've used this a few times before and like it.

The library [libwinmedia-py](https://github.com/mytja/libwinmedia-py) might be easier to get working on Windows, but it doesn't look like it is as well maintained as the others.

**Reply by srukale** [reply]
Hi @hx2A ,  we tried pydub library with ffmpeg dependency . The installation is not that difficult , just that it is not documented well at one place. So maybe while using pydub , we will make a good documentation so it becomes easier for users

**Reply by srukale** [reply]
Also @hx2A , I was suggesting the same, to have similar functionalities as Processing's sound library.

**Reply by hx2A** [reply]
> The installation is not that difficult 

That's encouraging! What platforms did you test the installation on? I would expect Linux to be easier than Windows or OSX.

**Reply by hx2A** [reply]
> I was suggesting the same, to have similar functionalities as Processing's sound library.

py5dub has good functionality, and can be a good foundation for this.

**Reply by srukale** [reply]
I tested on Windows, haven't tried OSX/Linux yet

**Reply by srukale** [reply]
> > I was suggesting the same, to have similar functionalities as Processing's sound library.
> 
> py5dub has good functionality, and can be a good foundation for this.

Okay , that sounds great then!

**Reply by hx2A** [reply]
> I tested on Windows, haven't tried OSX/Linux yet

Were you using Anaconda? Can you also try installing it with the Thonny editor?

**Comment by jaivanti** [comment]
I have thonny editor.....I will try running around with it

**Reply by srukale** [reply]
@hx2A I started with thonny only . Can we have a meet later some time, just to discuss what functionalities should we have in py5 sound library ? Till then we are exploring Librosa too , just to see whichever is comfortable to work with

**Reply by hx2A** [reply]
Sure, we can meet over zoom, but I don't have all the answers about what functionalities should be in the sound library. I would suggest implementing things in phases, with the first phase being the basics and then adding from there.

librosa is an excellent library also.

What time zone are you in? I am in EST. What is your availability next week?

**Reply by srukale** [reply]
Hi , we are in IST zone. How about someday 9pm in EST next week ? Yes, it will be logical to work in phases only .

**Reply by hx2A** [reply]
How about Tuesday 9 PM EST? Is that day Tuesday or Wednesday for you?

I'll need to send you a zoom link and I don't have your email address. You can get my email address at the bottom of http://ixora.io/

**Comment by jaivanti** [comment]
@hx2A So, sorry for the late reply.
Its 6.30 am Wednesday for us. We are fine with it.
This is our email ID:
jsdhokey_b18@pe.vjti.ac.in
srushtikale182@gmail.com

**Reply by hx2A** [reply]
OK, you should have gotten an email with the Zoom info. Talk to you later!

**Reply by srukale** [reply]
Yes, got the invite ! Thanks

**Comment by hx2A** [comment]
It was great speaking with you both yesterday!

Something I thought of afterwards is how a sound library could someday be a part of something that addresses py5's accessibility needs. Right now py5 is basically useless to someone using screen reader, and if it is possible to change that, I'd like to do so. I do care about accessibility but I don't know what to do here. I do know that such a thing would have to be done with guidance from experts on the subject and actual screen reader users, and that there's a lot more to it than "adding sound."

**Reply by srukale** [reply]
Accessibility in that case would be great and definitely mean something. The case is different to work on and also we don't have much idea about it . But while exploring sound options and researching about current objective, it would be great to see what people are working to address this issue. Just to confirm , accessibility here means to address for visually challenged people , right ?

**Reply by hx2A** [reply]
There's a lot I don't know about how to properly support folks who use screen readers or what they would want or expect from py5. I wanted to float the idea out there for you to be aware of as you work on the sound library. If I see an opportunity related to sound and accessibility, I'll let you know.

**Reply by srukale** [reply]
hi @hx2A , we were swamped up with our work last month . We had few ongoing things about sound library under process, will complete it and reach out to you soon. Sorry for not getting in contact last month.

**Reply by hx2A** [reply]
I understand, your job needs to be a priority. Good luck, and can't wait to see what you are working on!
